grant-project/grant-monorepo
# Create overhaul (Pt 2. - Team invites, user type consolidation)

Sorry, another big one fellas. Pretty sure this is the last one. Relies on #211. Fully closes #196, closes #146.

### What This Does

 The main star of the show is adding team members by email / address instead of manually making them, but there are a bunch of little improvements as well:

* Add team members to your draft proposal by email or address.
* Accept or reject invitations on your profile page.
* Refactors all user types to use a new and improved `User` type.
    * At some point, we'll want to take `emailAddress` away from this type and have a `FullUser` or something.
* Refactors social media to be broken out into `service`, `username`, and `url` fields.
* Adds a warning modal before publishing to tell you what's final, and optionally give you warnings (that you can ignore) about your proposal.
    * The only warning implemented right now is if you have invited team members who haven't accepted. But in the future, I can see us adding way more helpful hints, like adding more milestones, adding social medias to your team's accounts, and more.
* Refactors PUT /profile endpoint to have simpler inputs.
    

### Things to Improve

* Prevent users spamming invitations.
* Notifications for invitations. They're kind of tucked away now.
* Notifications to the proposal owner about invitation rejections / acceptances. 
* Clear out invitations for people after it's been published.

### Steps to Test

* Have 3 accounts available
* Create a new proposal
* Invite an account you have control of into that proposal by email address
* Invite the other account by ethereum address
* Attempt to publish the proposal as-is, confirm you get a warning
* Sign into both accounts and confirm you've gotten an invite on each
* Accept one and reject the other, confirm only one was added to the team
* Publish the proposal, confirm you no longer got warnings about unadded team members
* Confirm the team is correct

## Screenshots

### Invites

<img width="701" alt="screen shot 2018-11-19 at 4 01 57 pm" src="https://user-images.githubusercontent.com/649992/48734881-ce555500-ec14-11e8-8450-1f87495e5784.png">

### At review

<img width="388" alt="screen shot 2018-11-19 at 4 01 18 pm" src="https://user-images.githubusercontent.com/649992/48734902-d9a88080-ec14-11e8-852f-fc3835080c69.png">

### Invitations

<img width="932" alt="screen shot 2018-11-19 at 4 02 30 pm" src="https://user-images.githubusercontent.com/649992/48734905-e036f800-ec14-11e8-9d10-9d52a1d5c965.png">

### Warning before publish

<img width="538" alt="screen shot 2018-11-19 at 4 03 47 pm" src="https://user-images.githubusercontent.com/649992/48734918-e6c56f80-ec14-11e8-84ca-cf25b21e8969.png">
